base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -880,85 +880,125 @@ where B: BlockchainBackend
-        // validation can take some time during big blocks as it does Rangeproof and metadata signature validation.
-        // Because the sync process first acquires a read_lock then a write_lock, and the RWLock will be prioritised,
-        // the add_block write lock will be given out before the sync write_lock.
-        trace!(
-            target: LOG_TARGET,
-            "[add_block] waiting for write access to add block block #{}",
-            new_height
-        );
-        let timer = Instant::now();
-        let mut db = self.db_write_access()?;
+        let block_add_result = {
+            // This is important, we ask for a write lock to disable all read access to the db. The sync process sets
+            // the add_block disable flag,  but we can have a race condition between the two especially
+            // since the orphan validation can take some time during big blocks as it does Rangeproof and
+            // metadata signature validation. Because the sync process first acquires a read_lock then a
+            // write_lock, and the RWLock will be prioritised, the add_block write lock will be given out
+            // before the sync write_lock.
+            trace!(
+                target: LOG_TARGET,
+                "[add_block] waiting for write access to add block block #{} '{}'",
+                new_height,
+                block_hash.to_hex(),
+            );
 
-        trace!(
-            target: LOG_TARGET,
-            "[add_block] acquired write access db lock for block #{} in {:.2?}",
-            new_height,
-            timer.elapsed()
-        );
-        let block_hash = candidate_block.hash();
-        if db.contains(&DbKey::BlockHash(block_hash))? {
-            return Ok(BlockAddResult::BlockExists);
-        }
-        if db.bad_block_exists(block_hash)? {
-            return Err(ChainStorageError::ValidationError {
-                source: ValidationError::BadBlockFound {
-                    hash: block_hash.to_hex(),
-                },
-            });
-        }
-        // the only fast check we can perform that is slightly expensive to fake is a min difficulty check, this is done
-        // as soon as we receive the block before we do any processing on it. A proper proof of work is done as soon as
-        // we can link it to the main chain. Full block validation only happens when the proof of work is higher than
-        // the main chain and we want to add the block to the main chain.
-        let block_add_result = add_block(
-            &mut *db,
-            &self.config,
-            &self.consensus_manager,
-            &*self.validators.block,
-            &*self.validators.header,
-            self.consensus_manager.chain_strength_comparer(),
-            candidate_block,
-        )?;
+            let before_lock = timer.elapsed();
+            let mut db = self.db_write_access()?;
```

Comment:
You need to recheck for bad block and block exists here after getting the write lock